### DISCUSSION TITLE
UEFI installation always fail

### DISCUSSION THREAD

**Discussion by bobblkabb** [zbm-dev/zfsbootmenu#297] [Questions] [answered]
I have already worked through the [instructions](https://github.com/zbm-dev/zfsbootmenu/wiki/Void-Linux---Single-disk-UEFI) on 3 computers, point by point and always the encrypted version.
None of the installations worked.
At the point refind-install (I had chosen the pre-build zfsbootmenu variant) I got efi variable error messages. I then ran efivar-tester which claimed that the device was not UEFI capable.
The first two laptops have a CSM option in the bios, so I assumed that they are UEFI capable. And the third laptop (Chromebook Acer C720 with Coreboot) there is even a UEFI shell.
In this respect, I must have made a mistake when installing according to instructions. What could I have done wrong.

/sys/firmware/efi/efivars/ is empty

**Comment by Dragon-debian** [comment]
If `/sys/firmware/efi/efivars/` is empty, it means that `efivarfs` pseudo filesystem is not mounted there.The init system usually mounts this automatically.

Try to mount it manually and see what happens.
```bash
mount -t efivarfs efivarfs /sys/firmware/efi/efivars
```
you can replace second `efivarfs` with any string you want. its only a name.

If `mount` said ` unknown filesystem type 'efivarfs'`, the laptop most likely is booted in `legacy mode`.
check the `uefi settings` and make sure its booting in `uefi mode`, not `legacy support/mode` or `csm enabled`.

There are some cases that system is booted in `uefi mode` but you cant access EFI vars:

1- kernel does not support EFI. void linux kernel (and most other distros) has support for this. if you are using a custom kernel, check that kernel compiled with `CONFIG_EFI=y`.

2- you are using a real time kernel. efivarfs is disabled on those kernels by default. boot the kernel with `efi=runtime` kernel command line and try again.

3- `efivarfs` kernel module is not loaded. void linux kernel has this module builtin, so there is no need to run `modprobe efivarfs` if you are using void linux kernel.

4- kernel is booted with `noefi` kernel command line. if this is true, reboot and remove that from kernel command line.

**Comment by bobblkabb** [comment]
Yes, the scenario about the unknown filesystem happened.
I did another installation without UEFI because I wasn't sure about the UEFI capability of the laptop but it miserably failed.
So I installed once again - with UEFI. All your talking about the linux kernel reminded me that I probably took the latest kernel (5.15) not the kernel as instructed. Thought it was outdated. Now I get a proper UEFI Boot menu and I can unlock the hard disk but still having problems to load the system:
```
Last accessed by <unknown> (hostid=0) at Wed Apr 27 16:23:30 2022
The pool can be imported, use 'zpool import -f' to import the pool.
dracut Warning: ZFS: Unable to import pool zroot
cannot import 'zroot': pool was previously in use from another system.

dracut Warning: Can't mount root filesystem
```

**Comment by bobblkabb** [comment]
Changed my password like described [here](https://github.com/zbm-dev/zfsbootmenu/issues/159), I removed a special character:

1. Booted with hrmpf-Stick and chrooted into the system
2. echo 'my new password' > /etc/zfs/zroot.key
3. zfs change-key zroot
4. xbps-reconfigure -f linux5.10


Now I get following error during boot (after password prompt):
Wrong password
`zfs_mount_at() failed: encryption key not loadedEnrcypted ZFS password for zroot: Key load error: Failed to open key material file: No such file or directory`

**Comment by Dragon-debian** [comment, marked as answer]
Is key file avaiable in initramfs?
dracut should executes an emergency shell if it cant mount the root file system.
In that shell, check that `/etc/zfs/zroot.key` is available or not.
If its avaiable, check its content.
```
cat /etc/zfs/zroot.key
```

If the key file is ok, maybe `zpool.cache` is causing problem.
Boot ZFSBootMenu and boot into boot menu.
press `MOD + w` to reimport all the pools `rw`.
`MOD` can be `ctrl`, `alt`or `ctrl + alt`.

Jump into your boot environment with `MOD + j`.
Remove `zpool.cache` and create a new one.
```
rm /etc/zfs/zpool.cache
zpool set cachefile=/etc/zfs/zpool.cache zroot
```
EDIT: I added `--kver`.
Now, regenerate initramfs.
```
dracut -f --kver 5.15.34_1
```
Replace `5.15.34_1` with right kernel version. look in `/boot` to find the right one.

You can also reconfigure the kernel package to generate a new initramfs.
```
xbps-reconfigure -f linux5.15
```
Replace `linux5.15` with kernel version you have installed.

Both of the above commands can be used to regenerate initramfs in void linux. I dont know their difference.

**Comment by bobblkabb** [comment]
Deleting and setting works but fails at 'dracut -f':
```
dracut: Cannot find module directory /lib/modules/5.10.101_1/
dracut: and --no-kernel was not specified 
```
/lib/modules/ shows following folders:
5.10.111_1
5.15.34_1

So I tried 'dracut -f -kver /lib/modules/5.10.111_1':
```
dracut: -k/--kmoddir path must contain "lib/modules" as a parent of your kernel module directory,
dracut: or modules may not be placed int the correct location inside the initramfs.
dracut: was given: ver
dracut: expected: ./lib/modules/5.10.101_1
dracut: Please move your modules into the correct directory structure and pass the new location,
dracut: or set DRACUT_KMODDIR_OVERRIDE=1 to ignore the check.
```

So I copied 5.10.111_1 to 5.10.101_1 and tried it again with the same message as a result.

At last I succeeded with 'dracut -f --no-kernel', followed by 'xbps-reconfigure -f linux5.10'.
'...5.10.101_1 completed successful'-like message at end of the operation.

Now I get following error:
```
cannot import 'zroot': pool was previously in use from another system.
Last accessed by <unknown> (hostid=0)
The pool can be imported, use 'zpool import -f' to import the pool.
dracut Warning: ZFS: Unable to import pool zroot
```

I end up in the debug shell.

**Reply by Dragon-debian** [reply]
I forgot something. If you dont specify `--kver`for dracut, it guesses the right version from `uname -r` output.
The kernel version in ZFSBootMenu and your boot environment wasnt the same, so dracut failed to detect the kernel version and didnt generate an initramfs.

remove the copy you created, `/lib/modules/ 5.10.101_1`.

The right command that you had to execute was this:
```
dracut -f --kver 5.10.111_1
```
I think the `xbps-reconfigure` way is better as it passes the right kernel version to dracut and works fine in chroot.

**Comment by ahesford** [comment]
You don't have a consistent `/etc/hostid` in your ZBM and regular initramfs images. Either fix that or boot ZBM with the command-line options

    zbm.import_policy=hostid zbm.set_hostid

**Comment by bobblkabb** [comment]
I deleted the /lib/modules/5.10.101_1 folder again.

Then I entered chroot in read/write mode:
```
rm /etc/zfs/zpool.cache
zpool set cachefile=/etc/zfs/zpool.cache zroot
dracut -f --kver 5.10.111_1
```

It finishes with:
```
..
Creating image file '/boot/initramfs-5.10.111_1.img'
Creating initramfs image file '/boot/initramfs-5.10.111_1.img'
```

Now when I press [CTRL+K ] (kernels) and switch from 5.15.34_1 to 5.10.111_1 it works!
How do I make the 5.10.111_1 as default? Or how do I make 5.15.34_1 work?

**Reply by ahesford** [reply]
Either use the menu to set the default or read the zfsbootmenu.7 man page.

**Comment by bobblkabb** [comment]
Ha! Tricked the system. Just 'dracuted' the latest kernel and it works. Good a functional system.
Thank you very much for your support! Very appreciated.
